source/distributed/fullydistributed_tria.cc

```diff
@@ -0,0 +1,562 @@
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+#include <deal.II/base/logstream.h>
+#include <deal.II/base/memory_consumption.h>
+#include <deal.II/base/utilities.h>
+
+#include <deal.II/distributed/fullydistributed_tria.h>
+#include <deal.II/distributed/p4est_wrappers.h>
+
+#include <deal.II/grid/grid_tools.h>
+#include <deal.II/grid/tria.h>
+#include <deal.II/grid/tria_accessor.h>
+#include <deal.II/grid/tria_iterator.h>
+
+#include <deal.II/lac/dynamic_sparsity_pattern.h>
+#include <deal.II/lac/sparsity_tools.h>
+
+#include <algorithm>
+#include <fstream>
+#include <iostream>
+#include <numeric>
+#include <vector>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace parallel
+{
+  namespace fullydistributed
+  {
+    namespace internal
+    {
+      template <int dim>
+      bool
+      parent(typename CellId::binary_type c, typename CellId::binary_type p)
+      {
+        if (c[0] != p[0])
+          return false; // has same coarse cell
+        if ((c[1] >> 2) != ((p[1] >> 2) + 1))
+          return false; // level has one difference
+
+
+        const unsigned int n_child_indices = p[1] >> 2;
+        const unsigned int children_per_value =
+          sizeof(CellId::binary_type::value_type) * 8 / dim;
+        unsigned int child_level  = 0;
+        unsigned int binary_entry = 2;
+
+        // Loop until all child indices have been written
+        while (child_level < n_child_indices)
+          {
+            Assert(binary_entry < c.size(), ExcInternalError());
+
+            for (unsigned int j = 0; j < children_per_value; ++j)
+              {
+                if ((((c[binary_entry] >> (j * dim))) &
+                     (GeometryInfo<dim>::max_children_per_cell - 1)) !=
+                    (((p[binary_entry] >> (j * dim))) &
+                     (GeometryInfo<dim>::max_children_per_cell - 1)))
+                  return false;
+
+
+                ++child_level;
+                if (child_level == n_child_indices)
+                  break;
+              }
+
+            ++binary_entry;
+          }
+        return true;
+      }
+    } // namespace internal
+
+    template <int dim, int spacedim>
+    void
+    Triangulation<dim, spacedim>::reinit(
+      ConstructionData<dim, spacedim> &construction_data)
+    {
+      // check if there are locally relevant coarse-grid cells
+      if (construction_data.cells.size() == 0) // no
+        {
+          // 1) create a dummy hypercube
+          Point<spacedim> p1, p2;
+          for (unsigned int i = 0; i < dim; ++i)
```

Comment:
can you replace the whole code here by ``GridGenerator::hyper_cube(*this)``?

Reply:
I have done it with the help forward declaration (I really don't like forward declaration - it breaks clear dependencies).

Reply:
> I really don't like forward declaration - it breaks clear dependencies

What do you mean? I normally use forward declarations to be able to deal with dependencies properly. :slightly_smiling_face: